windows/win_nssm.ps1

```diff
@@ -0,0 +1,547 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+$ErrorActionPreference = "Stop"
+
+# WANT_JSON
+# POWERSHELL_COMMON
+
+$params = Parse-Args $args;
+$result = New-Object PSObject;
+Set-Attr $result "changed" $false;
+
+
+If ($params.name)
+{
+    $name = $params.name
+}
+Else
+{
+    Fail-Json $result "missing required argument: name"
+}
+
+If ($params.state)
+{
+    $state = $params.state.ToString().ToLower()
+    $validStates = "present", "absent", "started", "stopped", "restarted"
+    
+    # These don't really fit the declarative style of ansible
+    # If you need to do these things, you can just write a command for it
+    # "paused", "continued", "rotated"
+
+    If ($validStates -notcontains $state)
+    {
+        Fail-Json $result "state is $state; must be one of: $validStates"
+    }
+}
+else
+{
+    $state = "present"
+}
+
+If ($params.application)
+{
+    $application = $params.application
+}
+Else
+{
+    $application = $null
+}
+
+If ($params.app_parameters)
+{
+    $appParameters = $params.app_parameters
+}
+Else
+{
+    $appParameters = $null
+}
+
+If ($params.stdout_file)
+{
+    $stdoutFile = $params.stdout_file
+}
+Else
+{
+    $stdoutFile = $null
+}
+
+If ($params.stderr_file)
+{
+    $stderrFile = $params.stderr_file
+}
+Else
+{
+    $stderrFile = $null
+}
+
+Function Service-Exists
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name
+    )
+
+    return ,[bool](Get-Service "$name" -ErrorAction SilentlyContinue)
+}
+
+Function Nssm-Remove
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name
+    )
+
+    if (Service-Exists -name $name)
+    {
+        $cmd = "nssm stop ""$name"""
+        $results = invoke-expression $cmd
+
+        $cmd = "nssm remove ""$name"" confirm"
+        $results = invoke-expression $cmd
+
+        if ($LastExitCode -ne 0)
+        {
+            Set-Attr $result "nssm_error_cmd" $cmd
+            Set-Attr $result "nssm_error_log" "$results"
+            Throw "Error removing service ""$name"""
+        }
+
+        $result.changed = $true
+     }
+}
+
+Function Nssm-Install
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name,
+        [Parameter(Mandatory=$true)]
+        [string]$application
+    )
+
+    #note: the application name must look like the following, if the directory includes spaces:
+    # nssm install service "c:\Program Files\app.exe" """C:\Path with spaces"""
+    #see https://git.nssm.cc/?p=nssm/nssm.git;a=commit;h=0b386fc1984ab74ee59b7bed14b7e8f57212c22b for more info
+
+
+    if (!$application)
+    {
+        Throw "Error installing service ""$name"". No application was supplied."
+    }
+
+    if (!(Service-Exists -name $name))
```

Comment:
Currently, `Service-Exists` returns a collection with an element in it. Does the `!` operator do the intended thing there, if `@(true)` comes back? Unfamiliar with the operator unwrapping as I discovered from earlier reply.